Ngoc_Bao/myweb/app/Http/Controllers/Admin/ProductController.php

```diff
@@ -3,84 +3,215 @@
+        
+        $producthistorydetails = $producthistory->producthistorydetails->toArray();
 
-        return view('admin.products.show',['product'=>$product]);
+        return view('admin.products.show',[
+            'product'=>$product,
+            'user' => $user,
+            'producthistory' => $producthistory,
+            'producthistorydetails' => $producthistorydetails,
+        ]);
     }
 
     public function edit($id)
     {
-        $product = $this->productmodel->findOrFail($id);
-        $categories = $this->categorymodel->all();
+        $product = $this->modelProduct->findOrFail($id);
+        $categories = $this->modelCategory->all();
+        $user = \Auth::user();
 
         return view('admin.products.edit',[
             'product'=>$product,
             'categories'=>$categories,
+            'user' => $user
         ]);
     }
 
 
     public function update(Request $request, $id)
     {
-        
-        $product = $this->productmodel->findOrFail($id);
+        $product = $this->modelProduct->findOrFail($id);
+
+        $data = $request->only([
+            'category_id',
+            'name',
+            'description',
+            'price',
+            'quantity',
+            'sale_off',
+            'is_public',
+        ]);
 
-        $data = $request->all();
-        
-        $product['name'] = $data['name'];
-        $product['sale_off'] = $data['sale_off'];
-        $product['price'] = $data['price'];
-        $product['category_id'] = $data['category_id'];
-        $product['is_public'] = $data['is_public'];
-        $product['description'] = $data['description'];
-        $product->update();
+        $data['category_id'] = (int) $data['category_id'];
+        $data['is_public'] = isset($data['is_public']) ? (int) $data['is_public'] : 0;
+        $data['user_id'] = $product->id;
+
+        try {
+            $product->update($data);
+            $msg = 'Update product success.';
+
+
+            //  UPDATE PRODUCT HISTORY 
+            //
+            // find in table product_history collumn have user_id and product_id equal with $product->id and auth()->id
+            // if can't find,we have to create product_history then create product_history_Details
+
+
+            $producthistory = $this->modelProductHistory
+                                ->where('user_id', auth()->id())
+                                ->where('product_id', $product->id)
+                                ->get()
+                                ->toArray();
+            //dd($producthistory);
+
+            unset($data['user_id']);
+            if(count($producthistory) > 0) 
+            {
+
+                $data['product_history_id'] = $producthistory[0]['id'];
+                $producthistorydetail = $this->modelProductHistoryDetail->create($data);
+
+            } else 
+            {
+
+                $data_history_product['user_id'] = auth()->id();
```

Comment:
```suggestion
                $productHistoryData['user_id'] = auth()->id();
```
đặt biến kiểu camelCase nhé e